include/libhal-arm-mcu/stm32f1/adc.hpp

```diff
@@ -0,0 +1,157 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <libhal-arm-mcu/stm32f1/constants.hpp>
+#include <libhal/adc.hpp>
+#include <libhal/lock.hpp>
+#include <libhal/units.hpp>
+
+namespace hal::stm32f1 {
+
+/**
+ * @brief Manager for the stm32f1 series' onboard adc peripheral. Used to
+ * construct and manage channel objects that are tied to an adc. This also
+ * applies to the MCU variants that have more than one adc available.
+ *
+ */
+class adc_peripheral_manager final
+{
+public:
+  // Forward declaration.
+  class channel;
+  /**
+   * @brief Defines the pins which can be used for analog input for adc1 and
+   * adc2.
+   */
+  enum class pins : hal::u8
+  {
+    pa0 = 0,
+    pa1 = 1,
+    pa2 = 2,
+    pa3 = 3,
+    pa4 = 4,
+    pa5 = 5,
+    pa6 = 6,
+    pa7 = 7,
+    pb0 = 8,
+    pb1 = 9,
+    pc0 = 10,
+    pc1 = 11,
+    pc2 = 12,
+    pc3 = 13,
+    pc4 = 14,
+    pc5 = 15,
+  };
+
+  /**
+   * @brief Defines the available adc peripherals that CAN be onboard the MCU.
+   * Having two adc's provides the capability to do dual-channel conversions, as
+   * well as do simultaneous sampling through dual-mode (not currently
+   * supported). Note that the XL-density stm32f1 MCU's only have adc1.
+   *
+   */
+  enum class adc_selection : hal::u8
+  {
+    adc1 = 0,
+    adc2 = 1,
+  };
+
+  /**
+   * @brief Construct a new adc peripheral manager object.
+   *
+   * @param p_adc_selection - The specified adc peripheral to use.
+   * @param p_lock - An externally declared lock to use for thread safety when
+   * trying to read from the adc's. This can be a basic_lock or any type that
+   * derives from it.
+   */
+  adc_peripheral_manager(adc_selection p_adc_selection,
+                         hal::basic_lock& p_lock);
+
+  /**
+   * @brief Creates and configures an adc channel under the calling adc
+   * peripheral manager.
+   *
+   * @param p_pin - The pin to be used for the channels analog input.
+   * @return channel - is an implementation of the `hal::adc` interface that
+   * represents a channel on the ADC peripheral.
+   */
+  channel acquire_channel(pins p_pin);
+
+  adc_peripheral_manager(adc_peripheral_manager const& p_other) = delete;
+  adc_peripheral_manager& operator=(adc_peripheral_manager const& p_other) =
+    delete;
+  adc_peripheral_manager(adc_peripheral_manager&& p_other) noexcept = delete;
+  adc_peripheral_manager& operator=(adc_peripheral_manager&& p_other) noexcept =
+    delete;
+
+private:
+  /**
+   * @brief Takes an analog input reading.
+   *
+   * @param p_pin - The pin to read from.
+   * @return float - The sampled adc value.
+   */
```

Comment:
```suggestion
```
There is a comment below that explains why private stuff does not need to be doxygen commented.